### DISCUSSION TITLE
[Portenta H7] Threads/Cores

### DISCUSSION THREAD

**Discussion by darrenweinhold** [micropython/micropython#9777] [Hardware & Peripherals]
I am working on a project using the Portenta H7. 

Is dual core operation enabled? I noticed the Dual Flag set in the board config files [(here),](https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/mpconfigboard.mk) but didn't see any documentation on using this. I assume its just spawning two threads to run. If so, how does it specify which thread runs on which core?

**Comment by jimmo** [comment]
> Is dual core operation enabled? I noticed the Dual Flag set in the board config files [(here),](https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/mpconfigboard.mk) but didn't see any documentation on using this. I assume its just spawning two threads to run. If so, how does it specify which thread runs on which core?

MicroPython does not have any support for multi-core on chips where the cores are not identical. (i.e. we support configurations like the rp2 where the same code runs on both cores).

The comment in the config file is a bit misleading -- really what it's saying is that we're building firmware for the M7 core and therefore have the appropriate CMSIS configuration.

**Reply by darrenweinhold** [reply]
Got it. 

So assuming I wanted to run on both, what would be the best way to do that? Would that be to look at the rp2 codebase and adapt that? Or use RTOS to run the separate cores?

**Reply by jimmo** [reply]
> Would that be to look at the rp2 codebase and adapt that?

Maybe I misunderstood your question... why would the rp2 codebase help? You're using a H7 right?

> Or use RTOS to run the separate cores?

You have lots of options... but it probably depends a lot on exactly how you want this to work. e.g. always-on-m4 (low-power) turning on the high-performance core just for occasional tasks? Both running all the time? Do you need to run Micropython on both?

If you haven't already seen e.g. https://github.com/TannerGilbert/Arduino-Portenta-Examples/tree/master/dual-core this might give you some idea about how to approach this -- essentially you're writing two separate firmware images, and then given a mechanism for communication between them.

So for example as a simple starting point you could implement a "m4 worker" firmware in C/Arduino, and send commands to it from micropython. You would need to implement some python bindings for the intercore-communication which you could invoke from the python code running on the m7.

**Reply by darrenweinhold** [reply]
> > Would that be to look at the rp2 codebase and adapt that?
> 
> Maybe I misunderstood your question... why would the rp2 codebase help? You're using a H7 right?
> 

I was just thinking of an example of multiple cores running micropython. I haven't been able to find much beyond this.

This may be an example of an xy problem where I am aiming at the wrong solution. 
I have a use case where I am using 7 UARTs on the H7 to communicate with instruments and continually pull data. I also have a lot of computation that I want to do with the data to control some outputs in real time. The UART communication feels like a great fit for the M4 core and crunching data seems perfect for the M7 core. I'm not concerned about power; just ease of implementation, not missing cyclic data reads, and keeping the system responsive on the control side. Any thoughts are appreciated. 

In the meantime, I'll take a look at the example you provided. Thanks!
